homeassistant/auth/providers/__init__.py

```diff
@@ -228,18 +229,29 @@ def __init__(self, auth_provider: AuthProvider) -> None:
                     reason='login_expired'
                 )
 
-            result = await auth_module.async_validation(
+            result = await auth_module.async_validate(
                 self.user.id, user_input)  # type: ignore
             if not result:
                 errors['base'] = 'invalid_code'
+                self.invalid_mfa_times += 1
+                if (auth_module.MAX_RETRY_TIME > 0 and
+                        self.invalid_mfa_times >= auth_module.MAX_RETRY_TIME):
+                    return self.async_abort(
+                        reason='too_many_retry'
+                    )
 
             if not errors:
                 return await self.async_finish(self.user)
 
+        # MFA module may have init code need generate
```

Comment:
This can be done the first time the MFA form is requested

Reply:
Changed async_generate() to async_initialize()